File: src/alert.rs
/*  alerter: Alerter to chat servers
 *  Copyright (C) 2019 The alerter developers
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

pub mod alert_cli_parser;
pub mod config;
pub mod logging;
pub mod message;
pub mod util;

use crate::message::Level;
use crate::message::Message;

use std::collections::BTreeMap;
use std::io::Write;
use std::os::unix::net::UnixStream;

use chrono::Local;

use log::debug;
use log::error;
use log::warn;

fn main() {
    let arguments = alert_cli_parser::parse_arguments();
    logging::initialise(
        arguments
            .value_of(alert_cli_parser::FLAG_LOG_CONFIG)
            .expect("Missing default value in cli_parser"),
    );

    debug!("Starting up");

    let config_path = arguments
        .value_of(alert_cli_parser::FLAG_CONFIG)
        .expect("Missing default value in alert_cli_parser");
    debug!("Config is at {}", config_path);

    let config = config::parse_config(config_path);

    let message = compose_message_from_arguments(arguments);

    send_message(&config.socket_path, message);
}

fn compose_message_from_arguments(args: clap::ArgMatches) -> Message {
    Message {
        title: args
            .value_of(alert_cli_parser::FLAG_TITLE)
            .map(str::to_string)
            .unwrap_or_default(),

        text: args
            .value_of(alert_cli_parser::FLAG_TEXT)
            .map(str::to_string)
            .unwrap_or_default(),

        channel: args
            .value_of(alert_cli_parser::FLAG_CHANNEL)
            .map(str::to_string),

        level: args
            .value_of(alert_cli_parser::FLAG_LEVEL)
            .map(serde_yaml::from_str::<Level>)
            .map(Result::ok)
            .flatten()
            .unwrap_or_default(),

        link: args
            .value_of(alert_cli_parser::FLAG_TITLE_LINK)
            .map(str::to_string),

        version: env!("CARGO_BIN_NAME").to_string() + " v" + env!("CARGO_PKG_VERSION"),

        timestamp: Local::now(),

        fields: parse_additional_fields(args.values_of(alert_cli_parser::FLAG_FIELD)),
    }
}

fn parse_additional_fields(values: Option<clap::Values>) -> BTreeMap<String, String> {
    let mut fields = BTreeMap::default();
    if let Some(values) = values {
        for item in values {
            let mut split = item.splitn(2, ':');
            if let Some(key) = split.next() {
                if let Some(value) = split.next() {
                    fields.insert(key.to_string(), value.to_string());
                } else {
                    warn!("Skipping field '{}' because of missing ':'", item);
                }
            }
        }
    }
    fields
}

fn send_message(socket_path: &str, message: Message) {
    let mut stream = match UnixStream::connect(socket_path) {
        Err(e) => {
            error!("Failed to open socket: {}", e);
            return;
        }
        Ok(v) => v,
    };

    let raw_message = match serde_json::to_string(&message) {
        Err(e) => {
            error!("Failed to message to string: {}", e);
            return;
        }
        Ok(v) => v,
    };

    debug!("Sending {}", raw_message);

    if let Err(e) = stream.write_all(&raw_message.into_bytes()) {
        error!("Failed to hand over message to daemon: {}", e);
    }
}

File: test/src/test/java/de/njsm/alerter/test/client/Alert.java
/* alerter: Alerter to Slack
 * Copyright (C) 2019  The alerter developers
 *
 * This file is part of the alerter program suite.
 *
 * alerter is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * alerter is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package de.njsm.alerter.test.client;

import org.apache.commons.io.IOUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.Map;
import java.util.Optional;
import java.util.TreeMap;

public class Alert {

    private static final Logger LOG = LogManager.getLogger(Alert.class);

    public static final String BINARY_PATH = "de.njsm.alerter.binary";

    public static final String CONFIG_PATH = "de.njsm.alerter.config";

    public static final String LOG_CONFIG_PATH = "de.njsm.alerter.logconfig";

    private String text;

    private String title;

    private Optional<String> channel;

    private Optional<String> titleLink;

    private Optional<String> level;

    private final Map<String, String> fields;

    public Alert() {
        this.text = "";
        this.title = "";
        this.channel = Optional.empty();
        this.titleLink = Optional.empty();
        this.level = Optional.empty();
        this.fields = new TreeMap<>();
    }

    public static AlertBuilder build() {
        return new AlertBuilder();
    }

    private void call() {
        ArrayList<String> command = new ArrayList<>();

        command.add(getBinaryPath());
        command.add("-C");
        command.add(getConfigPath());
        command.add("-v");
        command.add(getLogConfigPath());

        command.add(title);
        command.add(text);

        channel.ifPresent(v -> {
            command.add("-c");
            command.add(v);
        });

        titleLink.ifPresent(v -> {
            command.add("-t");
            command.add(v);
        });

        level.ifPresent(v -> {
            command.add("-l");
            command.add(v);
        });

        fields.forEach((k,v) -> {
            command.add("-f");
            command.add(k + ":" + v);
        });

        try {
            Process p = Runtime.getRuntime().exec(command.toArray(new String[0]));
            p.waitFor();
            String stdout = IOUtils.toString(p.getInputStream(), Charset.defaultCharset());
            String stderr = IOUtils.toString(p.getErrorStream(), Charset.defaultCharset());
            LOG.debug("command: {}", command);
            if (!stdout.isEmpty())
                LOG.debug("stdout: {}", stdout);
            if (!stderr.isEmpty())
                LOG.debug("stderr: {}", stderr);
        } catch (InterruptedException | IOException e) {
            LOG.error("Failed to execute alert command", e);
        }
    }

    private String getBinaryPath() {
        return System.getProperty(BINARY_PATH);
    }

    private String getConfigPath() {
        return System.getProperty(CONFIG_PATH);
    }

    private String getLogConfigPath() {
        return System.getProperty(LOG_CONFIG_PATH);
    }

    public static class AlertBuilder {

        private final Alert product;

        public AlertBuilder() {
            product = new Alert();
        }

        public AlertBuilder withText(String text) {
            product.text = text;
            return this;
        }

        public AlertBuilder withTitle(String title) {
            product.title = title;
            return this;
        }

        public AlertBuilder withChannel(String channel) {
            product.channel = Optional.of(channel);
            return this;
        }

        public AlertBuilder withTitleLink(String titleLink) {
            product.titleLink = Optional.of(titleLink);
            return this;
        }

        public AlertBuilder withLevel(String level) {
            product.level = Optional.of(level);
            return this;
        }

        public AlertBuilder withField(String key, String value) {
            product.fields.put(key, value);
            return this;
        }

        public void call() {
            product.call();
        }
    }
}
